src/Development/IDE/Core/Shake.hs

```diff
@@ -538,6 +538,9 @@ usesWithStale key files = do
 
 withProgress :: (Eq a, Hashable a) => Var (HMap.HashMap a Int) -> a -> Action b -> Action b
 withProgress var file = actionBracket (f succ) (const $ f pred) . const
+    -- This functions are deliberately eta-expanded to avoid space leaks.
+    -- Do not remove the eta-expansion without profiling a session with at
+    -- least 1000 modifications.
```

Comment:
Once we are using `$!` and `evaluate` we can hang these comments off more solid guarantees.